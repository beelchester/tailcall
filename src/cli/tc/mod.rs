mod check;
mod gen;
mod helpers;
mod init;
pub mod run;
pub mod start;
