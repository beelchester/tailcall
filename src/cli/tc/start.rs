use std::sync::Arc;
use std::time::Duration;

use anyhow::{Context, Result};
use notify::{Config, EventKind, FsEventWatcher, RecommendedWatcher, RecursiveMode, Watcher};
use tokio::sync::broadcast;
use tokio::time::Instant;

use super::helpers::log_endpoint_set;
use crate::cli::fmt::Fmt;
use crate::cli::server::Server;
use crate::core::config::reader::ConfigReader;

pub(super) async fn start_command(
    file_paths: Vec<String>,
    watch: bool,
    config_reader: Arc<ConfigReader>,
) -> Result<()> {
    if watch {
        start_watch_server(file_paths, config_reader).await?;
    } else {
        let config_module = config_reader
            .read_all(&file_paths)
            .await
            .context("Failed to read config files")?;
        log_endpoint_set(&config_module.extensions().endpoint_set);
        Fmt::log_n_plus_one(false, config_module.config());
        let server = Server::new(config_module);
        server
            .fork_start(None)
            .await
            .context("Failed to start server")?;
    }
    Ok(())
}

/// Starts the server in watch mode
async fn start_watch_server(
    file_paths: Vec<String>,
    config_reader: Arc<ConfigReader>,
) -> Result<()> {
    let (tx, rx) = broadcast::channel(16);

    watch_files(&file_paths, tx, rx, config_reader).await?;

    Ok(())
}

/// Watches the file paths for changes
async fn watch_files(
    file_paths: &[String],
    tx: broadcast::Sender<()>,
    mut rx: broadcast::Receiver<()>,
    config_reader: Arc<ConfigReader>,
) -> Result<()> {
    let (watch_tx, watch_rx) = std::sync::mpsc::channel();
    // fake event to trigger the first server start
    watch_tx
        .clone()
        .send(Ok(notify::event::Event::new(
            notify::event::EventKind::Modify(notify::event::ModifyKind::Data(
                notify::event::DataChange::Content,
            )),
        )))
        .unwrap();
    let mut watcher = match RecommendedWatcher::new(watch_tx, Config::default()) {
        Ok(watcher) => watcher,
        Err(err) => {
            tracing::error!("Failed to create watcher: {}", err);
            return Ok(());
        }
    };

    for path in file_paths {
        if let Err(err) = watcher.watch(path.as_ref(), RecursiveMode::NonRecursive) {
            tracing::error!("Failed to watch path {:?}: {}", path, err);
        }
    }

    // Debounce delay to prevent multiple server restarts on a single file change
    // https://users.rust-lang.org/t/problem-with-notify-crate-v6-1/99877
    let debounce_duration = Duration::from_secs(1);
    let mut last_event_time = Instant::now() - debounce_duration;
    loop {
        match watch_rx.recv() {
            Ok(event) => {
                if let Ok(event) = event {
                    if let EventKind::Modify(notify::event::ModifyKind::Data(
                        notify::event::DataChange::Content,
                    )) = event.kind
                    {
                        let now = Instant::now();
                        if now.duration_since(last_event_time) >= debounce_duration {
                            last_event_time = now;

                            if let Err(err) = tx.send(()) {
                                tracing::error!("Failed to send the signal: {}", err);
                            }

                            if let Err(e) = handle_server(
                                &mut rx,
                                file_paths,
                                config_reader.clone(),
                                &mut watcher,
                            )
                            .await
                            {
                                tracing::error!("Failed to handle server: {}", e);
                            }
                        }
                    }
                }
            }
            Err(e) => tracing::error!("Watch error: {:?}", e),
        }
    }
}

/// Handles the server (in watch mode)
/// Prevents server crashes if config reader fails
async fn handle_server(
    rx: &mut broadcast::Receiver<()>,
    file_paths: &[String],
    config_reader: Arc<ConfigReader>,
    watcher: &mut FsEventWatcher,
) -> Result<()> {
    match config_reader.read_all(file_paths).await {
        Ok(config_module) => {
            let links = config_module
                .clone()
                .links
                .iter()
                .map(|link| link.src.clone())
                .collect::<Vec<_>>();
            for link in links {
                if let Err(err) = watcher.watch(link.as_ref(), RecursiveMode::NonRecursive) {
                    tracing::error!("Failed to watch path {:?}: {}", link, err);
                }
            }
            log_endpoint_set(&config_module.extensions().endpoint_set);
            Fmt::log_n_plus_one(false, config_module.config());

            let server = Server::new(config_module.clone());
            if let Err(err) = server.fork_start(Some(rx)).await {
                tracing::error!("Failed to start server: {}", err);
            }
        }
        Err(err) => {
            tracing::error!("Failed to read config files: {}", err);
        }
    }
    Ok(())
}
