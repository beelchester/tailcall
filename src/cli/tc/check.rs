use std::io::Write;

use anyhow::Result;

use super::helpers::{display_schema, log_endpoint_set};
use crate::cli::fmt::Fmt;
use crate::cli::CLIError;
use crate::core::blueprint::Blueprint;
use crate::core::config::reader::ConfigReader;
use crate::core::config::Source;
use crate::core::runtime::TargetRuntime;

pub struct CheckParams {
    pub file_paths: Vec<String>,
    pub n_plus_one_queries: bool,
    pub schema: bool,
    pub format: Option<Source>,
    pub runtime: TargetRuntime,
}

pub async fn check_command(
    params: CheckParams,
    config_reader: &ConfigReader,
    mut write_buf: Option<&mut dyn Write>,
) -> Result<()> {
    let CheckParams { file_paths, n_plus_one_queries, schema, format, runtime } = params;

    let config_module = (config_reader.read_all(&file_paths)).await?;
    log_endpoint_set(&config_module.extensions().endpoint_set);
    if let Some(format) = format {
        let format_msg = format.encode(&config_module)?;
        if let Some(write_buf) = &mut write_buf {
            writeln!(write_buf, "{}", format_msg)?;
        }
        Fmt::display(format_msg);
    }
    let blueprint = Blueprint::try_from(&config_module).map_err(CLIError::from);

    match blueprint {
        Ok(blueprint) => {
            let config_msg = format!("Config {} ... ok", file_paths.join(", "));
            let n_plus_message =
                Fmt::format_n_plus_one_message(n_plus_one_queries, config_module.config());

            if let Some(write_buf) = write_buf {
                writeln!(write_buf, "{}", config_msg)?;
                writeln!(write_buf, "{}", n_plus_message)?;
            }

            tracing::info!("{}", config_msg);
            tracing::info!("{}", n_plus_message);
            // Check the endpoints' schema
            let _ = config_module
                .extensions()
                .endpoint_set
                .clone()
                .into_checked(&blueprint, runtime)
                .await?;
            if schema {
                display_schema(&blueprint);
            }

            Ok(())
        }
        Err(e) => Err(e.into()),
    }
}
