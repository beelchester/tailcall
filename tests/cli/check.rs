pub mod test {
    use std::io::Cursor;

    use tailcall::cli::tc::check::{check_command, CheckParams};
    use tailcall::core::blueprint::Blueprint;
    use tokio::runtime::Runtime;

    pub fn run_check_command_spec(path: &std::path::Path) -> datatest_stable::Result<()> {
        let path = path.to_path_buf();
        let runtime = Runtime::new().unwrap();
        runtime.block_on(async move {
            run_test(&path.to_string_lossy()).await?;
            Ok(())
        })
    }

    async fn run_test(path: &str) -> anyhow::Result<()> {
        let runtime = tailcall::cli::runtime::init(&Blueprint::default());
        let config_reader = tailcall::core::config::reader::ConfigReader::init(runtime.clone());
        let params = CheckParams {
            file_paths: vec![path.to_string()],
            n_plus_one_queries: false,
            schema: false,
            format: None,
            runtime: runtime.clone(),
        };

        let mut output_buffer = Cursor::new(Vec::new());

        check_command(params, &config_reader, Some(&mut output_buffer)).await?;

        let output_string = String::from_utf8(output_buffer.into_inner())?;
        insta::assert_snapshot!(path, output_string);

        Ok(())
    }

    pub fn check_command_test(path: &std::path::Path) -> datatest_stable::Result<()> {
        let path = path.to_path_buf();
        let runtime = Runtime::new().unwrap();
        runtime.block_on(async move {
            run_test(&path.to_string_lossy()).await?;
            Ok(())
        })
    }
}

datatest_stable::harness!(
    test::check_command_test,
    "tests/cli/fixtures/check",
    r"^.*\..*$",
);
