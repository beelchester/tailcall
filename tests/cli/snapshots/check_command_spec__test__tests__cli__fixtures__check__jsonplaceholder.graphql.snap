---
source: tests/cli/check.rs
expression: output_string
---
Config tests/cli/fixtures/check/jsonplaceholder.graphql ... ok
N + 1 detected: 1
